dbms/src/Interpreters/joinDispatch.h

```diff
@@ -12,57 +12,50 @@
 namespace DB
 {
 
-template <bool fill_right, typename ASTTableJoin::Strictness>
-struct MapGetterImpl;
-
-template <>
-struct MapGetterImpl<false, ASTTableJoin::Strictness::Any>
-{
-    using Map = Join::MapsAny;
-};
-
-template <>
-struct MapGetterImpl<true, ASTTableJoin::Strictness::Any>
-{
-    using Map = Join::MapsAnyFull;
-};
-
-template <>
-struct MapGetterImpl<false, ASTTableJoin::Strictness::All>
-{
-    using Map = Join::MapsAll;
-};
-
-template <>
-struct MapGetterImpl<true, ASTTableJoin::Strictness::All>
-{
-    using Map = Join::MapsAllFull;
-};
-
-template <bool fill_right>
-struct MapGetterImpl<fill_right, ASTTableJoin::Strictness::Asof>
+template <ASTTableJoin::Kind kind, typename ASTTableJoin::Strictness>
+struct MapGetter;
+
+template <> struct MapGetter<ASTTableJoin::Kind::Left, ASTTableJoin::Strictness::RightAny> { using Map = Join::MapsOne; };
+template <> struct MapGetter<ASTTableJoin::Kind::Inner, ASTTableJoin::Strictness::RightAny> { using Map = Join::MapsOne; };
+template <> struct MapGetter<ASTTableJoin::Kind::Right, ASTTableJoin::Strictness::RightAny> { using Map = Join::MapsOneFlagged; };
+template <> struct MapGetter<ASTTableJoin::Kind::Full, ASTTableJoin::Strictness::RightAny> { using Map = Join::MapsOneFlagged; };
+
+template <> struct MapGetter<ASTTableJoin::Kind::Left, ASTTableJoin::Strictness::Any> { using Map = Join::MapsOne; };
+template <> struct MapGetter<ASTTableJoin::Kind::Inner, ASTTableJoin::Strictness::Any> { using Map = Join::MapsOneFlagged; };
+template <> struct MapGetter<ASTTableJoin::Kind::Right, ASTTableJoin::Strictness::Any> { using Map = Join::MapsAllFlagged; };
+template <> struct MapGetter<ASTTableJoin::Kind::Full, ASTTableJoin::Strictness::Any> { using Map = Join::MapsAllFlagged; };
+
+template <> struct MapGetter<ASTTableJoin::Kind::Left, ASTTableJoin::Strictness::All> { using Map = Join::MapsAll; };
+template <> struct MapGetter<ASTTableJoin::Kind::Inner, ASTTableJoin::Strictness::All> { using Map = Join::MapsAll; };
+template <> struct MapGetter<ASTTableJoin::Kind::Right, ASTTableJoin::Strictness::All> { using Map = Join::MapsAllFlagged; };
+template <> struct MapGetter<ASTTableJoin::Kind::Full, ASTTableJoin::Strictness::All> { using Map = Join::MapsAllFlagged; };
+
+/// Only SEMI LEFT and SEMI RIGHT are valid
+template <> struct MapGetter<ASTTableJoin::Kind::Left, ASTTableJoin::Strictness::Semi> { using Map = Join::MapsOne; };
+template <> struct MapGetter<ASTTableJoin::Kind::Inner, ASTTableJoin::Strictness::Semi> { using Map = Join::MapsOne; };
+template <> struct MapGetter<ASTTableJoin::Kind::Right, ASTTableJoin::Strictness::Semi> { using Map = Join::MapsAll; };
```

Comment:
Did not understand why not Join::MapsAllFlagged;
Am I right that `semi right` join should filter right table, so we probably need to add flag that right row was met?

Reply:
Yup, it's a bug